allennlp/training/callback_trainer.py

```diff
@@ -0,0 +1,250 @@
+        # I am not calling move_to_gpu here, because if the model is
+        # not already on the GPU then the optimizer is going to be wrong.
+        self.model = model
+        self.optimizer = optimizer
+        self.validate = False
+
+        # For capturing mid / end-of-epoch metrics
+        self.train_metrics: Dict[str, float] = {}
+        self.val_metrics: Dict[str, float] = {}
+        self.latest_val_metric = 0.0
+        self.train_loss = 0.0
+
+        # For capturing overall metrics
+        self.metrics: Dict[str, Any] = {}
+
+        self.batch_num_total = 0
+        self.batch_group: List[TensorDict] = []
+        self.batches_this_epoch = 0
+
+        self.training_batches: Iterable[List[TensorDict]] = ()
+        self.num_training_batches = 0
+
+        self.should_stop_early = False
+        self.num_epochs = num_epochs
+
+        self.training_start_time = 0.0
+        self.checkpoint_epoch: Union[int, str] = 0
+        self.model_save_interval = model_save_interval
+
+        self.last_log = 0.0
+        self.epoch_number = 0
+        self.batch_grad_norm: Optional[float] = None
+        self.handler = CallbackHandler(callbacks, self)
+
+    def batch_loss(self, batch_group: List[TensorDict], for_training: bool) -> torch.Tensor:
+        """
+        Does a forward pass on the given batches and returns the ``loss`` value in the result.
+        If ``for_training`` is `True` also applies regularization penalty.
+        """
+        if self._multiple_gpu:
+            output_dict = training_util.data_parallel(batch_group, self.model, self._cuda_devices)
+        else:
+            assert len(batch_group) == 1
+            batch = batch_group[0]
+            batch = nn_util.move_to_device(batch, self._cuda_devices[0])
+            output_dict = self.model(**batch)
+
+        try:
+            loss = output_dict["loss"]
+            if for_training:
+                loss += self.model.get_regularization_penalty()
+        except KeyError:
+            if for_training:
+                raise RuntimeError("The model you are trying to optimize does not contain a"
+                                   " 'loss' key in the output of model.forward(inputs).")
+            loss = None
+
+        return loss
+
+    def train(self) -> Dict[str, Any]:
+        """
+        Trains the supplied model with the supplied parameters.
+        """
+        self.handler.fire_event(Events.RESTORE_CHECKPOINT)
+        starting_epoch = self.epoch_number
+
+        logger.info("Beginning training.")
+        self.handler.fire_event(Events.TRAINING_START)
+
+        self.training_start_time = time.time()
+
+        for self.epoch_number in range(starting_epoch, self.num_epochs):
+            epoch_start_time = time.time()
+            ####
+            self.handler.fire_event(Events.EPOCH_START)
+
+            self.train_loss = 0.0
+            # Set the model to "train" mode.
+            self.model.train()
```

Comment:
I feel like this should be in the `Events.EPOCH_START` handler.  Context: I'm looking for where things would break if you make a `GanTrainer` that does `super().__init__(model=None, optimizer=None)`.  I think this is the only thing that needs to be fixed, and then my suggestion should just work.